packages/three-vrm/src/VRMUtils/removeUnnecessaryJoints.ts

```diff
@@ -46,64 +46,72 @@ export function removeUnnecessaryJoints(
   let maxBones = 0;
 
-  // Iterate over all skinned meshes and collect bones and boneInverses
+  // Iterate over all skinned meshes and remap bones for each skin index attribute
   for (const mesh of skinnedMeshes) {
     const geometry = mesh.geometry;
     const attribute = geometry.getAttribute('skinIndex');
 
-    const bones: THREE.Bone[] = []; // new list of bone
-    const boneInverses: THREE.Matrix4[] = []; // new list of boneInverse
-    const boneIndexMap: { [index: number]: number } = {}; // map of old bone index vs. new bone index
+    if (attributeToBoneIndexMapMap.has(attribute)) {
+      continue;
+    }
+
+    const oldToNew = new Map<number, number>(); // map of old bone index vs. new bone index
+    const newToOld = new Map<number, number>(); // map of new bone index vs. old bone index
 
     // create a new bone map
     for (let i = 0; i < attribute.count; i++) {
       for (let j = 0; j < attribute.itemSize; j++) {
-        const index = attribute.getComponent(i, j);
+        const oldIndex = attribute.getComponent(i, j);
+        let newIndex = oldToNew.get(oldIndex);
 
         // new skinIndex buffer
-        if (boneIndexMap[index] == null) {
-          boneIndexMap[index] = bones.length;
-          bones.push(mesh.skeleton.bones[index]);
-          boneInverses.push(mesh.skeleton.boneInverses[index]);
+        if (newIndex == null) {
+          newIndex = oldToNew.size;
+          oldToNew.set(oldIndex, newIndex);
+          newToOld.set(newIndex, oldIndex);
         }
 
-        attribute.setComponent(i, j, boneIndexMap[index]);
+        attribute.setComponent(i, j, newIndex);
       }
     }
 
     // replace with new indices
     attribute.needsUpdate = true;
 
     // update boneList
-    bonesList.set(mesh, { bones, boneInverses });
+    attributeToBoneIndexMapMap.set(attribute, newToOld);
 
     // update max bones count
-    maxBones = Math.max(maxBones, bones.length);
+    maxBones = Math.max(maxBones, oldToNew.size);
   }
 
   // Let's actually set the skeletons
   for (const mesh of skinnedMeshes) {
-    const { bones, boneInverses } = bonesList.get(mesh)!;
+    const geometry = mesh.geometry;
+    const attribute = geometry.getAttribute('skinIndex');
+    const newToOld = attributeToBoneIndexMapMap.get(attribute)!;
+
+    const bones: THREE.Bone[] = [];
+    const boneInverses: THREE.Matrix4[] = [];
 
     // if `experimentalSameBoneCounts` is `true`, compensate skeletons with dummy bones to keep the bone count same between skeletons
-    if (experimentalSameBoneCounts) {
-      for (let i = bones.length; i < maxBones; i++) {
-        bones[i] = bones[0];
-        boneInverses[i] = boneInverses[0];
-      }
+    const nBones = experimentalSameBoneCounts ? maxBones : newToOld.size;
+
+    for (let newIndex = 0; newIndex < nBones; newIndex++) {
+      const oldIndex = newToOld.get(newIndex) ?? 0;
+
+      bones.push(mesh.skeleton.bones[oldIndex]);
+      boneInverses.push(mesh.skeleton.boneInverses[oldIndex]);
```

Comment:
Considering cases multiple meshes share the same skin index attribute but not the same bones (which is unlikely) or boneInverses (which is more likely), we moved these lines here